### PR TITLE
Bugfix/subnets duplicate

### PR DESCRIPTION
**Pull request by sayboras**
### Description

Fixed https://github.com/weaveworks/eksctl/issues/1468

### Checklist
- [x] Added tests that cover your change (if possible)
- [ ] Added/modified documentation as required (such as the `README.md`, and `examples` directory)
- [ ] Manually tested

### CODE REVIEW

File: pkg/apis/eksctl.io/v1alpha5/types.go [hunk 1]
```diff
@@ -2,6 +2,7 @@ package v1alpha5
 
 import (
 	"fmt"
+	"k8s.io/kops/util/pkg/slice"
```
Comment (sayboras) [inline pkg/apis/eksctl.io/v1alpha5/types.go]: ops 😢

**Comment by sayboras** [comment]
@errordeveloper @martina-if @cPu1 for your review and feedback.

File: pkg/apis/eksctl.io/v1alpha5/types.go [hunk 2]
```diff
@@ -421,6 +423,21 @@ func (c *ClusterConfig) AppendAvailabilityZone(newAZ string) {
 	c.AvailabilityZones = append(c.AvailabilityZones, newAZ)
 }
 
+// CleanupSubnets clean up subnet entries having invalid AZ
+func (c *ClusterConfig) CleanupSubnets() {
```
Comment (martina-if) [inline pkg/apis/eksctl.io/v1alpha5/types.go]: Since this is a very specific thing for subnets and it's an implementation detail I think I would move it to the vpc.go file, moving the receiver to an argument.

**Comment by martina-if** [comment]
@sayboras since this was a regression, can you add some unit tests so that this doesn't happen again?

**Comment by sayboras** [comment]
> @sayboras since this was a regression, can you add some unit tests so that this doesn't happen again?

Sure, I will address the lack of coverage soon. Please let me know if you have other comments as well.

**Comment by sayboras** [comment]
@martina-if the test cases for `vpc.go` is very limited, and probably wrongly assert the result. I am slowly adding the test cases to improve coverage. This will take times to actually reach the point for the actual fixes.

File: pkg/vpc/vpc_test.go [hunk 3]
```diff
@@ -91,4 +102,82 @@ var _ = Describe("VPC Endpoints", func() {
 			Error:       errors.New(eks.ErrCodeResourceNotFoundException),
 		}),
 	)
+
+	DescribeTable("can import VPC",
```
Comment (martina-if) [inline pkg/vpc/vpc_test.go]: Can we add a test that checks that given an input that would normally produce duplicate subnets checks that there are no dups?

**Comment by martina-if** [comment]
One more nitpick: can you rewrite the commit message so it says something relevant to the change instead of "rebased from master"?

File: pkg/vpc/vpc_test.go (hunk 3 above)
Comment (sayboras) [reply pkg/vpc/vpc_test.go]: Agreed that this test case must be added. However, all the methods in this file is public package methods, so it's hard to mock the entire method stack. 

So just want to check with you if I can refactor the code a little bit with dependency injection for private method like this commit https://github.com/weaveworks/eksctl/pull/1477/commits/1905d0b667c118b352943e629d0b253008b78cd3

File: pkg/apis/eksctl.io/v1alpha5/types.go (hunk 2 above)
Comment (martina-if) [reply pkg/apis/eksctl.io/v1alpha5/types.go]: @sayboras I think I prefer the way the tests were before. What I was looking for was something like this:
```
// For corner cases test closer to the function
testCleanup() {
  c = ClusterConfig{ .... }  // test data
  CleanupSubnets(c)
  Expect(c.subnets).To(//not have duplicates)
}

// One regression test for the public API that makes sure we don't have this issue again
testImport() {
// Mock EC2 calls
  p.MockEC2().On("DescribeVpcsInput", MatchedBy(func(input *ec2.DescribeVpcsInput) bool {
      return input != nil
  })).Return(...)

  ImportAllSubnets(p, spec) // Use the mocked provider
  
  Expect(spec).To(// not have duplicates...)
}
```
What do you think? At the top of the tests for "VPC - Cluster Endpoints" you can see an example on how we use the EC2 mocks.

Comment (sayboras) [reply pkg/apis/eksctl.io/v1alpha5/types.go]: It's great idea, I should have thought about it earlier. Maybe I just want to understand and test the whole thing, which makes my mind so narrowed :cry:

File: pkg/vpc/vpc.go [hunk 4]
```diff
@@ -137,11 +139,11 @@ func UseFromCluster(provider api.ClusterProvider, stack *cfn.Stack, spec *api.Cl
 	return outputs.Collect(*stack, requiredCollectors, optionalCollectors)
 }
 
-// Import will update spec with VPC ID/CIDR
+// importVPC will update spec with VPC ID/CIDR
```
Comment (sayboras) [inline pkg/vpc/vpc.go]: @martina-if this function is used in the same package only, so making private

File: pkg/vpc/vpc.go [hunk 5]
```diff
@@ -74,7 +76,7 @@ func describeSubnets(provider api.ClusterProvider, subnetIDs ...string) ([]*ec2.
 	return output.Subnets, nil
 }
 
-func describe(provider api.ClusterProvider, vpcID string) (*ec2.Vpc, error) {
+func describeVPC(provider api.ClusterProvider, vpcID string) (*ec2.Vpc, error) {
```
Comment (sayboras) [inline pkg/vpc/vpc.go]: `describe` might be just a little bit too general, so I renamed to describe VPC.

**Comment by martina-if** [comment]
Thanks @sayboras !
